pkg/controller/catalogsource/catalogsource_controller.go

```diff
@@ -0,0 +1,138 @@
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+// Add creates a new CatalogSource Controller and adds it to the Manager. The Manager will set fields on the Controller
+// and Start it when the Manager is Started.
+func Add(mgr manager.Manager, _ options.ControllerOptions) error {
+	return add(mgr, newReconciler(mgr))
+}
+
+// newReconciler returns a new reconcile.Reconciler
+func newReconciler(mgr manager.Manager) reconcile.Reconciler {
+	client := mgr.GetClient()
+	return &ReconcileCatalogSource{
+		client: client,
+	}
+}
+
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+
+	c, err := controller.New("catalogsource-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	_, defaultCatalogsources := defaults.GetGlobalDefinitions()
+	pred := predicate.Funcs{
+		CreateFunc: func(e event.CreateEvent) bool {
+			return false
+		},
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			if _, ok := defaultCatalogsources[e.MetaOld.GetName()]; ok {
+				return true
+			}
+			return false
+		},
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			if _, ok := defaultCatalogsources[e.Meta.GetName()]; ok {
+				// If DeleteStateUnknown is true it implies that the Delete event was missed
+				// and we can ignore it.
+				if e.DeleteStateUnknown {
+					return false
+				}
+				return true
+			}
+			return false
+		},
+		GenericFunc: func(e event.GenericEvent) bool {
+			if _, ok := defaultCatalogsources[e.Meta.GetName()]; ok {
+				return true
+			}
+			return false
+		},
+	}
+
+	err = c.Watch(&source.Kind{Type: &olm.CatalogSource{}}, &handler.EnqueueRequestForObject{}, pred)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// blank assignment to verify that ReconcileOperatorHub implements reconcile.Reconciler
+var _ reconcile.Reconciler = &ReconcileCatalogSource{}
+
+// ReconcileCatalogSource reconciles a CatalogSource object
+type ReconcileCatalogSource struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver
+	client client.Client
+}
+
+func (r *ReconcileCatalogSource) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	log.Infof("Reconciling default CatalogSource %s", request.Name)
+
+	_, defaultCatalogsources := defaults.GetGlobalDefinitions()
+	defaultCatsrcDef := defaultCatalogsources[request.Name]
+	// Fetch the CatalogSource instance
+	instance := &olm.CatalogSource{}
+	err := r.client.Get(context.TODO(), request.NamespacedName, instance)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			if !operatorhub.GetSingleton().Get()[defaultCatsrcDef.Name] {
+				createNewCatsrcInstance(r.client, defaultCatsrcDef)
+			}
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+
+	if instance.DeletionTimestamp != nil {
```

Comment:
I see what the check here is but it seems like a race against the GC controller. If its marked for deletion, but not gone yet, recreate it. I wonder if this is a necessary as it would only apply in the case where it was deleted right after the `Get()`, otherwise you would be recreating an object still on the cluster. 

But maybe other controllers use this pattern, I'm not sure.

Reply:
GC = garbage collector

Reply:
@exdx I put this in because the reconcile request could be received before the object is actually removed (even though it has been marked for deletion). The comments [here](https://github.com/operator-framework/operator-marketplace/blob/release-4.5/pkg/controller/operatorsource/operatorsource_controller.go#L89-L108) might make that clearer. 

If we don't put in this clause, a sequence of event might look like this: 
1) Object has been marked for deletion 
2) Reconciler sends a get request, and gets the object, checks if the spec has been tampered with, and sees nothing wrong and does nothing 
3) Object is deleted (and no reconciliation request is sent again coz it has already been sent)

Reply:
I see the logic you're describing, makes sense. But I'm concerned about the case where the object is marked for deletion, we issue the get request on the next line and it succeeds, and _then_ the GC controller deletes it immediately afterwards (so the `createNew()...` doesn't get called). Basically a race against when the GC controller actually deletes the object. 

Maybe its safer to requeue in the case where its marked for deletion?

Reply:
I see what you mean. Update to re-queue the request after 5 secs instead of calling `createNewCatsrcInstance()`